### DISCUSSION TITLE
Error code 5 (Connection Refused - Not Authorized) for serverless broker over TLS with CA cert with nRF5340 board

### DISCUSSION THREAD

**Discussion by mayli10** [emqx/emqx#13547] [Q&A]
Hello! I am using the nRF5340 dev board with an nRF7002ek wifi shield to connect to MQTT using the [Zephyr MQTT library](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/mqtt.h) and MQTT 3.1.1 (the only supported version by Zephyr). I am connecting with TLS over MQTT, not web sockets TLS/SSL and building this firmware in C.

I have a serverless EMQX deployment (free tier, barely used) and I'm attempting to connect my dev board to MQTT over TLS secure connection. 

My current approach:
- I first connect to wifi and set a 30 second timeout to let wifi fully initialize
- I then use the connection details provided by my EMQX dashboard to initialize the client and connect to MQTT
- the details I use are the ones include the broker hostname: {omitted}.ala.us-east-1.emqxsl.com, the port 8883, and the CA certificate I downloaded directly from my EMQX dashboard
- I also included my username and password that I set in my EMQX dashboard
- here are the additional configs I set when I attempt to connect that I am suspicious of:
	tls_cfg->peer_verify = TLS_PEER_VERIFY_OPTIONAL;
	tls_cfg->cipher_count = 0;
	tls_cfg->cipher_list = NULL;
	tls_cfg->sec_tag_count = ARRAY_SIZE(sec_tag_list);
	tls_cfg->sec_tag_list = sec_tag_list;
	tls_cfg->session_cache = TLS_SESSION_CACHE_DISABLED;
	tls_cfg->hostname = {omitted}.ala.us-east-1.emqxsl.com; (should I be setting a host name or just leaving it as NULL)?
- Additionally, my keep-alive value for my MQTT client is 60 seconds and am creating a clean session (CONFIG_MQTT_CLEAN_SESSION=y) each time I connect to MQTT
- I also am running out of memory on my dev board (running at 97% FLASH and 94% RAM) so I am also worried this may be a problem

The problem:
- I am able to successfully connect to MQTT with my current connection details, but then I hit the CONNACK case and get error code 5 (Connection Refused - Not Authorized)
- **I have confirmed that this is only an issue I have with EMQX and the way I'm setting up my TLS config** because I am able to connect and send/receive messages to the public EMQX server without TLS and I also have been able to securely connect with TLS using the test.mosquitto.org broker along with their provided CA cert
- I have also tested using my EMQX connection details with my EMQX CA certificate and username/password on the MQTTX desktop app and the MQTT Explorer desktop app and was able to connect through there.

_I would really appreciate any help or ideas because I've been wrangling this issue for a while now and would be eternally grateful for any suggestions!! Thank you in advance :)_

Here are the serial monitor error logs I am seeing:
```
wifi connect -s "{omitted wifi name}" -k 1 -p "{omitted wifi password}"
Connection requested
Connected
[00:00:40.750,183] <inf>: Added CA certificate, size: 1339 bytes
[00:00:40.750,183] <inf>: Initializing MQTT client
[00:00:40.851,776] <inf>: IPv4 address of MQTT broker found {omitted ipv4 address}
[00:00:40.851,806] <inf>: Successfully initialized broker connection
[00:00:40.852,172] <inf>: Now establishing a connection to the MQTT broker
[00:00:41.988,433] <inf>: Successfully established a connection to the MQTT broker
[00:00:42.068,237] <inf>: CA certificate size: 1339
[00:00:42.068,237] <err>: MQTT connect failed: 5, return code: 5
[00:00:42.076,721] <inf>: MQTT client disconnected: -111
[00:00:42.076,721] <err>: Error in mqtt_input: -111
[00:00:42.076,721] <inf>: Disconnecting MQTT client
[00:00:42.076,751] <err>: Could not disconnect MQTT client: -128
[00:00:42.076,782] <inf>: Reconnecting in 60 seconds...
```

**Comment by savonarola** [comment]
It's difficult to localize such a problem neither having the code nor the logs. An option to investigate may be:

* run EMQX v5 locally, e.g. in as a docker image with self-signed certs;
* try to reproduce the problem and get the logs.

**Comment by zibuyu1995** [comment]
Hi there,

Thank you for providing detailed information about your connection issue.

Given that the Serverless tier uses a multi-tenant architecture, it is crucial for your client to pass the Server Name Indication (SNI) during the TLS handshake. Without the SNI, EMQX Cloud Serverless cannot identify the tenant and will refuse the connection, which is likely causing the "Connection Refused - Not Authorized" error you are experiencing.

Please ensure that you are setting the SNI correctly in your TLS configuration. You should set the `tls_cfg->hostname` to your broker hostname `{omitted}.ala.us-east-1.emqxsl.com`.

Additionally, if you continue to experience issues, I recommend submitting a support ticket through the [EMQX Cloud Console](https://cloud-intl.emqx.com/console/tickets). Our professional technical support team will be able to provide you with more detailed assistance.

Best regards,  
EMQX Cloud Support Team

**Comment by OpenDTB** [comment]
Same question with you. I try to use BC20&STM32 to connect emqx Serverless via mqtts,but it's easy to open,and hard to connect,finally,I got a code5

**Reply by zmstone** [reply]
Hi  @OpenDTB 

Could you confirm that you have set SNI ?

As in the previous reply @mayli10 
> You should set the tls_cfg->hostname to your broker hostname {omitted}.ala.us-east-1.emqxsl.com

If issue persists, you are welcome to open a support ticket [EMQX Cloud Console](https://cloud-intl.emqx.com/console/tickets)
Providing your client information such as client ID, username, or source IP etc,  so the team can run some traces to find the cause.

**Comment by Channel59** [comment]
In Zephyr, SNI is not enabled by default. Even if you set the hostname property, the SNI extension is not compiled and connecting to the EMQX Serverless broker will fail.

Enable by adding 
```
CONFIG_MBEDTLS_SERVER_NAME_INDICATION=y
```
to your proj.conf file.
